keps/sig-api-machinery/3920-finish-feature-flags/README.md

```diff
@@ -0,0 +1,951 @@
+* Expose the feature states in use in the cluster
+* Make it possible to get the cluster back into the "off" state
+* Expose the "off but dangling references / uses" state
+* Make it easy to add enablement and disablement code
+* Make it easy to add upgrade and downgrade code
+* Make it easy to tell if it is not safe to upgrade / downgrade due to feature
+  states
+* A test ensures that feature on -> use -> feature off works for all features
+  (i.e. it's easy for feature authors to be included in this test)
+
+Secondary goals:
+* Third parties should be able to use the API and libraries we build for their
+  own features.
+* Set feature states centrally (on apiserver command line) -- may not apply to
+  all features
+* Permit the platform admin to permit selected features to be dynamically
+  settable, by api clients
+
+Follow-up goals:
+* When a feature is about a new field, remove it from OpenAPI specs when the
+  feature is off.
+* For features that possibly could support it, add a canarying mechanism so that
+  the feature enablement doesn't break the whole cluster at once when there is a
+  problem.
+
+### Non-Goals
+
+* The feature system will not permit passing configuration, since that's already
+  done in a variety of ways today. Therefore only features that don't take
+  config can be dynamic. Maybe this can be addressed in a follow up.
+
+## Proposal
+
+### High-level design
+
+First we will discuss all the pieces with minimal detail. Probably this section
+will leave you with questions which hopefully are answered in the "detailed
+design" section.
+
+#### Client Flow
+
+Let's call the current use of feature gates the "client flow". Things that are
+already handled correctly by the existing client flow we will try not to change.
+
+The main thing we want to add to the client flow is reporting of the feature's
+uses: did the client check the feature state, and what value did it find? The
+necessity of this will be clear when we discuss the "server flow" later.
+
+To make our improvements to the client flow, we will add a new implementation of
+the existing k8s.io/component-base/featuregate package's MutableFeatureGate
+type. It will require post-construction initialization, so that it can write the
+uses to the api; to be a drop-in replacement, it can't require an api client
+when it is constructed, because that happens before the command line is parsed /
+config files are read.
+
+Our new client will use the existing MutableFeatureGate to read the command
+line, so perhaps it should be thought of as a layer.
+
+When we update this KEP to solve the secondary goals, we will add here a
+mechanism for reading feature on/off values from the API -- that's not easy, due
+to the startup order.
+
+#### API
+
+Obviously, we will add an API object, Feature (details below). Features will be
+namespaced; this will permit all Kubernetes features to be in the kube-system
+namespace, and third parties can use another namespace.
```

Comment:
I love that this can be used by third parties, that is really nice. What roles will have write access to the Feature API? 

As I read this, it sounds like you are saying: anyone (with the right permissions) can add features in a namespace, and then this allows third parties to use this feature API in their own namespace to manage features of their component. Upstream k8s components will put their features in the kube-system namespace.

That's a reasonable approach. However, here's a crazy thought. Could we instead create a ClusterFeature and Feature, ala ClusterRole and Role, and allow features to be enabled and disabled on a per-namespace basis?

Whether it is an upstream feature or a third-party feature would just be an issue of the name of the Feature, and namespace would be irrelevant. On a per-namespace basis, a feature could be used to override the global state (if allowed). Something like:

ClusterFeature:
 - on: on cluster wide, cannot be overridden per namespace
 - default-on: on cluster wide, can be overridden per namespace
 - off: off cluster wide, cannot be overridden per namespace
 - default-off: off cluster wide, can be overridden per namespace

Feature of the same name can then be defined in a namespace as on/off, and of course status will reflect the composition of that and ClusterFeature.

If the ClusterFeature does not exist, then it is assumed to be "default-off". This allows third-parties to manage their namespaced features without a cluster or platform admin adding the ClusterFeature.

Reply:
Obviously for some features, only cluster-wide makes sense. But this could be a nice way to safely try features that only have impact on individual resources or can otherwise be namespace scoped.

Reply:
I do want to add canarying capabilities to this eventually, but not all features can be on/off on a namespace basis. I was going to try and avoid discussing how to make canarying work until we can agree on the rest of it :)

I was already planning on using namespaces to permission different classes of features differently, I don't think I should also use them for canarying in this way-- I have other plans that I'd like to keep vague until later :)

Reply:
Ok, sounds good. I do think it's a step beyond "canarying", in the sense that namespaced feature enablement actually enables better multi-tenancy in the cluster (one less thing everyone has to agree on). I think what I describe above enables the same functionality the use of namespaces in the KEP does, plus more. But perhaps it conflicts with your thinking on the canarying, I look forward to seeing that.